[build] guard against transient kokoro build failures

Example:

```
> Configure project :
flutterPluginVersion: 94.0.0-dev.20260629-75eb6ab
androidStudioVersion: 2025.3.4.6
dartPluginVersion: 505.0.0
sinceBuild: 251
untilBuild: 261.*
javaVersion: 21
group: io.flutter

FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':compileJava'.
> Could not resolve all dependencies for configuration ':compileClasspath'.
   > java.nio.file.ClosedFileSystemException (no error message)

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights from a Build Scan (powered by Develocity).
> Get more help at [https://help.gradle.org](https://www.google.com/url?q=https://help.gradle.org&sa=D).

BUILD FAILED in 4m 8s


[ID: 4045508] Command finished after 393 secs, exit value: 1

```

(Follow-up from https://github.com/flutter/flutter-intellij/pull/9010.)

--- a/tool/kokoro/build.sh
+++ b/tool/kokoro/build.sh
@@ -4,36 +4,6 @@ setup
 
 echo "kokoro build start"
 
-MAX_RETRIES=3
-RETRY_DELAY_SECS=15
-ATTEMPT=1
-
-# Retry the build a few times to address transient failures.
-# See: https://github.com/flutter/flutter-intellij/issues/9009
-while [[ $ATTEMPT -le $MAX_RETRIES ]]; do
-  echo "Running buildPlugin (Attempt $ATTEMPT of $MAX_RETRIES)..."
-  
-  if ./gradlew buildPlugin; then
-    echo "Gradle build completed successfully."
-    break
-  else
-    exit_code=$?
-    if [[ $exit_code -ge 128 ]]; then
-      echo "Build interrupted. Exiting."
-      exit $exit_code
-    fi
-  fi
-
-  echo "Build failed on attempt $ATTEMPT."
-  
-  if [[ $ATTEMPT -eq $MAX_RETRIES ]]; then
-    echo "All $MAX_RETRIES attempts failed. Exiting."
-    exit $exit_code
-  fi
-
-  echo "Waiting $RETRY_DELAY_SECS seconds before retrying..."
-  sleep $RETRY_DELAY_SECS
-  ATTEMPT=$((ATTEMPT + 1))
-done
+run_gradle_with_retry buildPlugin
 
 echo "kokoro build finished"

--- a/tool/kokoro/deploy.sh
+++ b/tool/kokoro/deploy.sh
@@ -8,7 +8,7 @@ setup
 
 echo "kokoro build start"
 
-./gradlew buildPlugin -Pdev --no-configuration-cache
+run_gradle_with_retry buildPlugin -Pdev --no-configuration-cache
 
 echo "kokoro build finished"
 

--- a/tool/kokoro/setup.sh
+++ b/tool/kokoro/setup.sh
@@ -1,9 +1,83 @@
 #!/bin/bash
 
+# Run a gradle command with retries to handle transient errors.
+# See:
+# - https://github.com/flutter/flutter-intellij/issues/9009
+# - https://github.com/flutter/flutter-intellij/issues/9021
+# Usage: run_gradle_with_retry [--max-retries N] [--delay-secs N] <gradle_args...>
+run_gradle_with_retry() {
+  local max_retries=2
+  local delay_secs=15
+
+  while [[ $# -gt 0 ]]; do
+    case "$1" in
+      --max-retries)
+        if [[ $# -lt 2 ]]; then
+          echo "Error: --max-retries requires an argument." >&2
+          return 1
+        fi
+        max_retries="$2"
+        shift 2
+        ;;
+      --delay-secs)
+        if [[ $# -lt 2 ]]; then
+          echo "Error: --delay-secs requires an argument." >&2
+          return 1
+        fi
+        delay_secs="$2"
+        shift 2
+        ;;
+      *)
+        break
+        ;;
+    esac
+  done
+
+  local gradle_args=("$@")
+  local total_attempts=$((1 + ${max_retries:-0}))
+  local ATTEMPT=1
+  local exit_code=0
+
+  local gradlew_cmd="./gradlew"
+  if [[ ! -x "$gradlew_cmd" ]]; then
+    gradlew_cmd="./third_party/gradlew"
+  fi
+
+  echo "Gradle retry config: max_retries=$max_retries, total_attempts=$total_attempts, delay_secs=$delay_secs" >&2
+
+  while [[ $ATTEMPT -le $total_attempts ]]; do
+    echo "Running $gradlew_cmd ${gradle_args[*]} (Attempt $ATTEMPT of $total_attempts)..." >&2
+    
+    if "$gradlew_cmd" "${gradle_args[@]}"; then
+      echo "Gradle command completed successfully." >&2
+      return 0
+    else
+      exit_code=$?
+      if [[ $exit_code -ge 128 ]]; then
+        echo "Gradle command interrupted. Exiting." >&2
+        return $exit_code
+      fi
+    fi
+
+    echo "Gradle command failed on attempt $ATTEMPT." >&2
+    
+    if [[ $ATTEMPT -eq $total_attempts ]]; then
+      echo "All $total_attempts attempts failed." >&2
+      return $exit_code
+    fi
+
+    echo "Waiting $delay_secs seconds before retrying..." >&2
+    sleep $delay_secs
+    ATTEMPT=$((ATTEMPT + 1))
+  done
+}
+
 # Initialize everything required by the plugin tool.
 setup() {
   # Fail on any error.
   set -e
+  # Prevent pipeline failures from being masked (e.g. printVersion | tail).
+  set -o pipefail
 
   # Display commands being run. Only do this while debugging and be careful
   # that no confidential information is displayed.
@@ -45,5 +119,5 @@ setup() {
   export NO_FS_ROOTS_ACCESS_CHECK=true
 
   (cd tool/plugin; echo "dart pub get `pwd`"; dart pub get --no-precompile)
-  ./gradlew --version
+  run_gradle_with_retry --version
 }

--- a/tool/kokoro/setup.sh
+++ b/tool/kokoro/setup.sh
@@ -16,12 +16,20 @@ run_gradle_with_retry() {
           echo "Error: --max-retries requires an argument." >&2
           return 1
         fi
+        if [[ ! "$2" =~ ^[0-9]+$ ]]; then
+          echo "Error: --max-retries requires a non-negative integer." >&2
+          return 1
+        fi
         max_retries="$2"
         shift 2
         ;;
       --delay-secs)
         if [[ $# -lt 2 ]]; then
           echo "Error: --delay-secs requires an argument." >&2
+          return 1
+        fi
+        if [[ ! "$2" =~ ^[0-9]+$ ]]; then
+          echo "Error: --delay-secs requires a non-negative integer." >&2
           return 1
         fi
         delay_secs="$2"

--- a/tool/kokoro/test.sh
+++ b/tool/kokoro/test.sh
@@ -6,6 +6,6 @@ setup
 (cd testData/sample_tests; echo "dart pub get `pwd`"; dart pub get --no-precompile)
 
 echo "kokoro test start"
-./gradlew test
+run_gradle_with_retry --max-retries 0 test
 
 echo "kokoro test finished"

